### PR TITLE
Support SkipLocalsinit

### PR DESCRIPTION
**Pull request by kerams**
Implements https://github.com/fsharp/fslang-suggestions/issues/990.

When applied on a module, class or method/function, all nested methods/functions and lambdas must not set the `.locals init` flag (paraphrased based on [docs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.skiplocalsinitattribute?view=net-5.0)).

This naive implementation assumes that locals should be initialized by default unless stated otherwise, though that [may change in the future](https://github.com/fsharp/fslang-suggestions/issues/990#issuecomment-800166484). It seems to work well at first glance with the exception of using the attribute on a class. The `eenvinner` in `GenTypeDef` correctly recognizes that locals should not be initialized, but the env instance is supposedly not passed down to `GenMethodForBinding` for the class' instance and static methods, unlike when the attribute is used on an entire module. Any clues?

Please let me know if what I did makes sense. I'll also need some pointers on authoring tests for this kind of thing.

### CODE REVIEW

**Comment by TIHan** [comment]
Looks straight forward so far. But yea you are right, the `eenvinner` with the `initLocals=true` isn't getting passed down. Will have to look at a bit more closely at this.

**Comment by kerams** [comment]
Do I just create a test file similar to https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Compiler.ComponentTests/EmittedIL/TailCalls.fs or do I have to muck with stuff like this https://github.com/dotnet/fsharp/tree/main/tests/fsharpqa/Source/CodeGen/EmittedIL?

**Comment by vzarytovskii** [comment]
> Do I just create a test file similar to https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Compiler.ComponentTests/EmittedIL/TailCalls.fs or do I have to muck around with stuff like this https://github.com/dotnet/fsharp/tree/main/tests/fsharpqa/Source/CodeGen/EmittedIL?

Creating a test in ComponentTests suite should be fine.

**Comment by kerams** [comment]
@vzarytovskii, the problem here is the fact that the attribute is .NET 5+ only, so the test snippet needs to be compiled against that TFM. What now?

**Comment by kerams** [comment]
It looks like the test only failed during the net472 run, so one option is running the test behind something like `#if NET_5`, even though there is no reason why it shouldn't run on net472 with an explicit target framework.

**Comment by kerams** [comment]
`--targetprofile:netcore` does not guarantee consistent results across platforms either. I'm afraid being able to specify the TFM is the only option left.

File: src/fsharp/TcGlobals.fs [hunk 1]
```diff
@@ -1192,6 +1192,7 @@ type public TcGlobals(compilingFslib: bool, ilg:ILGlobals, fslibCcu: CcuThunk, d
   member val attrib_CallerLineNumberAttribute = findSysAttrib "System.Runtime.CompilerServices.CallerLineNumberAttribute"
   member val attrib_CallerFilePathAttribute = findSysAttrib "System.Runtime.CompilerServices.CallerFilePathAttribute"
   member val attrib_CallerMemberNameAttribute = findSysAttrib "System.Runtime.CompilerServices.CallerMemberNameAttribute"
+  member val attrib_SkipLocalsInitAttribute  = findSysAttrib "System.Runtime.CompilerServices.SkipLocalsInitAttribute"
```
Comment (kerams) [inline src/fsharp/TcGlobals.fs:1195]: I'm not quite sure if this shouldn't be `tryFindSysAttrib`. Attributes above use both find and try find and I can't tell what the deciding factor is.

**Comment by kerams** [comment]
@TIHan, got it. Methods don't get the class's `eenvinner` because they are processed _after_, not inside `GenTypeDef`, so every method has to individually check its as well as the parent class's attributes.

https://github.com/dotnet/fsharp/blob/04ef0674f9338b7987f0f6a568bac632515cf46b/src/fsharp/IlxGen.fs#L6894-L6901

Tests pass when run on against .NET 5, so this is ready for review. I don't know what to do about 'Build Windows desktop_release'.

**Comment by kerams** [comment]
I'd appreciate help with that test.

**Comment by TIHan** [comment]
Apologies for the delay. I will find time today to look at this. I've been doing similar stuff in IlxGen lately.

File: src/fsharp/TcGlobals.fs (hunk 1 above)
Comment (TIHan) [reply src/fsharp/TcGlobals.fs:1195]: This is fine. You can check to see if you can deref it. If you can't, then it couldn't find it.

**Review by TIHan — APPROVED** [review]
This looks great! You don't have to worry about checking for a deref on the attribute; `HasFSharpAttribute` will return false if the `SkipLocalsInitAttribute` could not be found in the framework.

In regards to the test, you can put a `#if NETCOREAPP` around the test so it will only run on netcore and not framework.

**Comment by kerams** [comment]
Thanks, hopefully it's green and ready to go now.

**Review by cartermp — APPROVED** [review]
I think this looks good, but should it also be backed by LangVersion? It's not really a new feature per say, just respecting an existing attribute. So I say it's not necessary. But I'm happy to hear arguments otherwise.

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/SkipLocalsInit.fs [hunk 2]
```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace FSharp.Compiler.ComponentTests.EmittedIL
+
+open Xunit
+open FSharp.Test.Utilities.Compiler
+
+#if NETCOREAPP
```
Comment (vzarytovskii) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/SkipLocalsInit.fs:8]: That should be good testing wise, I'll see what we can do with the x-framework IL validation.

**Review by vzarytovskii — APPROVED** [review]
Thanks!

**Comment by kerams** [comment]
On the off chance that you're waiting to hear my opinion, I don't think LangVersion is needed either. The attribute is fairly new and, more importantly, currently has no impact in F#, so I find it extremely unlikely there's code out there that uses it. The risks (someone using `SkipLocalsInit`, but actually relying on initialized locals, recompiling their code and suddenly getting different runtime behavior) are thus negligible.

**Comment by cartermp** [comment]
Agreed. Let's merge it!

**Comment by dsyme** [comment]
There is a problem here

We use locals zero init to generate default values of arbitrary value types here:

```
            let ilTy = GenType cenv.amap m eenv.tyenv ty
            LocalScope "ilzero" cgbuf (fun scopeMarks ->
                let locIdx, realloc, _ =
                    // Ensure that we have an g.CompilerGlobalState
                    assert(g.CompilerGlobalState |> Option.isSome)
                    AllocLocal cenv cgbuf eenv true (g.CompilerGlobalState.Value.IlxGenNiceNameGenerator.FreshCompilerGeneratedName ("default", m), ilTy, false) scopeMarks
                // "initobj" (Generated by EmitInitLocal) doesn't work on byref types
                // But ilzero(&ty) only gets generated in the built-in get-address function so
                // we can just rely on zeroinit of all IL locals.
                if realloc then
                    match ilTy with
                    | ILType.Byref _ -> ()
                    | _ -> EmitInitLocal cgbuf ilTy locIdx
                EmitGetLocal cgbuf ilTy locIdx
            )
```

You should change the `if realloc` to be `if realloc or <skipping locals init>`.

However that leaves a separate problem about how to generate a zero byref value, because from the comment `initobj` evidently can't be used with byref type.   I think if this code is generated for the `ILType.Byref` case we should still set the zero init flag.  I'm not entirely sure what code would generate a zero byref value (you can't do `Unchecked.defaultof<byref<int>>`) but it may be possible somehow

**Comment by kerams** [comment]
@dsyme

This code

```fsharp
[<System.Runtime.CompilerServices.SkipLocalsInit>]
let z () =
    let mutable a = Unchecked.defaultof<System.DateTime>
    a <- Unchecked.defaultof<System.DateTime>
```

generates

```
.locals (
	[0] valuetype [System.Runtime]System.DateTime,
	[1] valuetype [System.Runtime]System.DateTime
)

/* 0x0000025C 07           */ IL_0000: ldloc.1
/* 0x0000025D 0A           */ IL_0001: stloc.0
/* 0x0000025E 1201         */ IL_0002: ldloca.s  V_1
/* 0x00000260 FE1503000001 */ IL_0004: initobj   [System.Runtime]System.DateTime
/* 0x00000266 07           */ IL_000A: ldloc.1
/* 0x00000267 0A           */ IL_000B: stloc.0
/* 0x00000268 2A           */ IL_000C: ret
```

With `if realloc && cenv.initLocals then` I get

```
.locals (
	[0] valuetype [System.Runtime]System.DateTime,
	[1] valuetype [System.Runtime]System.DateTime
)

/* 0x0000025C 07           */ IL_0000: ldloc.1
/* 0x0000025D 0A           */ IL_0001: stloc.0
/* 0x0000025E 07           */ IL_0002: ldloc.1
/* 0x0000025F 0A           */ IL_0003: stloc.0
/* 0x00000260 2A           */ IL_0004: ret
```

Is that what you want to see? I'm not sure I follow your second point. If the compiler didn't generate init code for byrefs before, why would `SkipLocalsInitAttribute` make a difference?

**Comment by kerams** [comment]
Ah, perhaps you want to always zero init when locals are no longer initialized, so `Unchecked.defaultof` would not return the default but an undefined/invalid state in the case of structs?

**Comment by dsyme** [comment]
I think you want `if realloc || not eenv.initLocals`

**Comment by kerams** [comment]
Ok, will send a PR with a test for this.

**Pull request merged**